src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,538 @@
+        /// </summary>
+        /// <param name="outerChannel"></param>
+        /// <returns></returns>
+        internal IServiceChannelDispatcher GetInnerServiceChannelDispatcher(IReplyChannel outerChannel)
+        {
+           Task<IServiceChannelDispatcher> channelTask = InnerServiceDispatcher.CreateServiceChannelDispatcherAsync(outerChannel);
+           innerServiceChanelDispatcher = channelTask.GetAwaiter().GetResult();
+            return innerServiceChanelDispatcher;
+        }
+
+        //Reference OnAcceptChannel/SecurityChannelListner
+        private IServiceChannelDispatcher GetInnerChannelDispatcher(IChannel outerChannel)
+        {
+            IServiceChannelDispatcher securityChannelDispatcher = null;
+            SecurityProtocol securityProtocol = this.SecurityProtocolFactory.CreateSecurityProtocol(null, null,
+            (outerChannel is IReplyChannel || outerChannel is IReplySessionChannel), TimeSpan.Zero);
+            securityProtocol.Open(TimeSpan.Zero);
+            /* TODO once we add more features
+            if (outerChannel is IInputChannel)
+            {
+                securityChannel = new SecurityInputChannel(listener, (IInputChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IInputSessionChannel))
+            {
+                securityChannel = new SecurityInputSessionChannel(listener, (IInputSessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IDuplexChannel))
+            {
+                securityChannel = new SecurityDuplexChannel(listener, (IDuplexChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else if (outerChannel is IDuplexSessionChannel))
+            {
+                securityChannel = new SecurityDuplexSessionChannel(listener, (IDuplexSessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+            }
+            else*/
+            if (outerChannel is IReplyChannel)
+            {
+                securityChannelDispatcher = new SecurityReplyChannelDispatcher(this, (IReplyChannel)outerChannel, securityProtocol, this.settingsLifetimeManager);
+            }
+            /* else if (listener.SupportsRequestReply && typeof(TChannel) == typeof(IReplySessionChannel))
+             {
+                 securityChannel = new SecurityReplySessionChannel(listener, (IReplySessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+             }
+             else
+             {
+                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.GetString(SR.UnsupportedChannelInterfaceType, typeof(TChannel))));
+             }*/
+
+            return securityChannelDispatcher;
+        }
+    }
+
+    internal abstract class ServerSecurityChannel<UChannel> where UChannel : class //, IChannel
```

Comment:
As all the types which derive from this class are IServiceChannelDispatcher's, this should be renamed to ServerSecurityChannelDispatcher and implement IServiceChannelDispatcher. The DispatchAsync methods then should be declared on this class but marked as abstract. Then you can remove the implements interface on the derived class and mark the DispatchAsync methods as override's.